js/bitopro.js

```diff
@@ -0,0 +1,1349 @@
+        //                 "lastPrice":"21.26110000",
+        //                 "isBuyer":false,
+        //                 "priceChange24hr":"-6.53",
+        //                 "volume24hr":"102846.47084802",
+        //                 "high24hr":"23.24460000",
+        //                 "low24hr":"21.13730000"
+        //             }
+        //         ]
+        //     }
+        //
+        return this.parseTickers (tickers, symbols);
+    }
+
+    async fetchOrderBook (symbol, limit = undefined, params = {}) {
+        const timestamp = this.milliseconds ();
+        await this.loadMarkets ();
+        const request = {
+            'pair': this.marketId (symbol),
+        };
+        if (limit !== undefined) {
+            request['limit'] = limit;
+        }
+        const response = await this.publicGetOrderBookPair (this.extend (request, params));
+        //
+        //     {
+        //         "bids":[
+        //             {
+        //                 "price":"1175271",
+        //                 "amount":"0.00022804",
+        //                 "count":1,
+        //                 "total":"0.00022804"
+        //             }
+        //         ],
+        //         "asks":[
+        //             {
+        //                 "price":"1176906",
+        //                 "amount":"0.0496",
+        //                 "count":1,
+        //                 "total":"0.0496"
+        //             }
+        //         ]
+        //     }
+        //
+        return this.parseOrderBook (response, symbol, timestamp, 'bids', 'asks', 'price', 'amount');
+    }
+
+    parseTrade (trade, market) {
+        //
+        // fetchTrades
+        //         {
+        //             "timestamp":1644651458,
+        //             "price":"1180785.00000000",
+        //             "amount":"0.00020000",
+        //             "isBuyer":false
+        //         }
+        //
+        // fetchMyTrades
+        //         {
+        //             "id":"8384438099",
+        //             "pair":"bnb_twd",
+        //             "price":"16000",
+        //             "avgExecutionPrice":"0",
+        //             "action":"SELL",
+        //             "type":"LIMIT",
+        //             "timestamp":1644902118265,
+        //             "status":4,
+        //             "originalAmount":"0.01",
+        //             "remainingAmount":"0.01",
+        //             "executedAmount":"0",
+        //             "fee":"0",
+        //             "feeSymbol":"twd",
+        //             "bitoFee":"0",
+        //             "total":"0",
+        //             "seq":"BNBTWD8984230567",
+        //             "updatedTimestamp":1644902118265
+        //         }
+        //
+        const id = this.safeString (trade, 'id');
+        let timestamp = this.safeInteger (trade, 'timestamp');
+        if (id === undefined) {
+            timestamp = this.safeTimestamp (trade, 'timestamp');
```

Comment:
We don't need to do both safeInteger and safeTimestamp for a public trade returned from `fetchTrades()`. Instead, let's do one of the two, but not both, depending on the presence of the `id`:

```JavaScript
let timestamp = undefined;
if (id === undefined) {
    timestamp = this.safeTimestamp (trade, 'timestamp');
} else {
    timestamp = this.safeInteger (trade, 'timestamp');
}
```